lib/concurrent/async.rb

```diff
@@ -42,12 +125,35 @@ def validate_argc(obj, method, *args)
         raise ArgumentError.new("wrong number of arguments (#{argc} for #{arity}..*)")
       end
     end
-    module_function :validate_argc
+
+    # @!visibility private
+    def self.included(base)
+      base.extend(ClassMethods)
+      base.send(:private_class_method, :new)
+      super(base)
+    end
+
+    # @!visibility private
+    def self.extended(base)
+      base.extend(ClassMethods)
+      base.send(:private_class_method, :new)
+      super(base)
+    end
+
+    # @!visibility private
+    module ClassMethods
+      def create(*args, &block)
+        obj = self.send(:new, *args, &block)
+        obj.send(:init_synchronization)
```

Comment:
if user calls one of the async methods in initialize (even that it's a bad practice) this ca fail. Swapping the order: `allocate`, `init_synchronization`, `initialize` should fix it.